src/main/src/main/java/org/geoserver/security/impl/DefaultResourceAccessManager.java

```diff
@@ -493,12 +506,26 @@ private String[] getLayerGroupPath(LayerGroupInfo layerGroup) {
 
     @Override
     public Filter getSecurityFilter(Authentication user, Class<? extends CatalogInfo> clazz) {
+        Filter filter = filterCache.getIfPresent(Pair.of(user, clazz));
+        if (filter == null) {
+            filter = getSecurityFilterInternal(user, clazz);
+            if (filter != null) {
+                filterCache.put(Pair.of(user, clazz), filter);
+            }
+        }
+        if (filter == null) {
+            return InMemorySecurityFilter.buildUserAccessFilter(this, user);
+        } else {
+            return filter;
+        }
+    }
+
+    public Filter getSecurityFilterInternal(Authentication user, Class<? extends CatalogInfo> clazz) {
         if (getMode() == CatalogMode.CHALLENGE) {
             // If we're in CHALLENGE mode, we cannot pre-filter
             // for the other types we have no clue, use the in memory filtering
-            return InMemorySecurityFilter.buildUserAccessFilter(this, user);
+            return null;
```

Comment:
The change makes the comment above harder to understand... are we actually moving to "in memory filtering" here. Not really, it's returning null because in the case of "challenge" mode we don't want to cache the filter, right? But I'm not sure why.... just because this filter is not expensive to build?

Reply:
I honestly find it confusing. IMO it calls for a couple test cases too, besides improving the comment for accuracy

Reply:
see my other comment, I wanted to avoid putting the InMemorySecurityFilter in the cache. I will make improvements